Add characters count in the table of contents

Jackie6 opened the pull request:
## Description
fix #13796 

> When writing a post in Chinese the word count shown in the content structure does not show an accurate word count.

## Screenshot
![image](https://user-images.githubusercontent.com/24849874/54852493-65194780-4cc3-11e9-8af2-7195ff820b8f.png)


## How has this been tested?
- [X] Local
- [X] Browser testing

## Types of changes
Add the count of characters(no spaces) in the table of contents

## Testing

Steps to reproduce the behavior:
1. Use http://generator.lorem-ipsum.info/_chinese to generate some sample Chinese text
2. Create new page in Gutenberg
3. Paste in content
4. Click the info icon at the top
5. See if the number of characters is correct

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [X] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [X] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [X] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

Jackie6 committed:
Add characters count in the table of contents

kjellr commented:
With the addition of this item, we'll also want to adjust the spacing for the `table-of-contents__count` items. Currently, it looks a little weird to have that last "Blocks" item sitting off on its own:

 
![Screen Shot 2019-04-02 at 4 12 25 PM](https://user-images.githubusercontent.com/1202812/55432953-37a18900-5562-11e9-93e6-9fa0971060d0.png)


Two options: 

1. Widen the modal, remove [the existing `25%` width rule](https://github.com/WordPress/gutenberg/blob/94f54892c214ebd014ec126a9575b7e902198ad9/packages/editor/src/components/table-of-contents/style.scss#L26), and give each items some right-padding instead: 

![Screen Shot 2019-04-02 at 4 08 17 PM](https://user-images.githubusercontent.com/1202812/55432748-af22e880-5561-11e9-993f-4b6c8f5152e3.png)

2. Change the `25%` width rule to `33.33%` instead, to force the 4th and 5th items onto the next line. We'll also want to add a tiny bit of bottom padding to these items. This keeps things compact, but also avoids having just one item all alone: 

![Screen Shot 2019-04-02 at 4 01 42 PM](https://user-images.githubusercontent.com/1202812/55432845-f315ed80-5561-11e9-927a-f61405d5e21f.png)

Option 2 seems like the simplest fix to me.

Jackie6 commented:
Hi, @kjellr Thanks for your comments. I think the option two looks better for me. But we may still need to wait for some design feedback.

talldan commented on packages/editor/src/components/char-count/index.js:
```diff
@@ -0,0 +1,24 @@
+	const charCountType = _x( 'characters_excluding_spaces', 'Word count type. Do not translate!' );
+
+	return (
+		<span className="word-count">{ wordCount( content, charCountType ) }</span>
```
`word-count` isn't the right class name here. Instead I think it should be called `char-count`, reflecting the component name. To get the styling to work an extra selector will have to be added here for the new `.char-count` class:
https://github.com/WordPress/gutenberg/blob/4fc05d35b710e4545c4a07e04d2f1e2bad692373/packages/editor/src/components/table-of-contents/style.scss#L33-L39

It is a bit unusual that these components (`CharCount`, `WordCount`) receive their style from a completely different component (`TableOfContentsPanel`), but I think it's outside of the scope of this PR to fix that.

talldan commented on packages/editor/src/components/char-count/index.js:
```diff
@@ -0,0 +1,24 @@
+ */
+import { withSelect } from '@wordpress/data';
+import { _x } from '@wordpress/i18n';
+import { count as wordCount } from '@wordpress/wordcount';
```
The `wordCount` alias is a bit misleading, given it returns that characterCount when used in this file. It's minor, but it'd be good to either leave it as `count` or alias it as `charCount`.

talldan reviewed (commented):
Hey @Jackie6, this looks good, thanks so much for your contribution. I did some manual testing, and the character count looks good. I tried a few edge cases, like characters that are represented by multiple code points, and the character count looks correct.

I've noticed a couple of minor things with the code, if you're able to tackle those and the discussed style changes it'd be much appreciated.

talldan commented:
Hey @Jackie6, just thought I'd note that the feedback from @kjellr was the design feedback.

Your PR was discussed during the design triage meeting:
(https://wordpress.slack.com/archives/C02S78ZAL/p1554221572033000 - you may have to sign up to slack to view the discussion).

It seems like you both agree on option 2, so that looks like a good way forward.

Jackie6 committed:
Merge branch 'update/word-count' of https://github.com/Jackie6/gutenberg into update/word-count

Jackie6 committed:
Change the class of char-count and add new css class

Jackie6 committed:
Make the table-of-contents have three coloums

Jackie6 commented:
Hi, @talldan thank you very much for your efforts and time in this pull request. I have updated the related files according to your suggestions. I have one more question. Do we need to update the user manual or other documentation about this change? I mean we have added a new item in the table of contents which is the number of characters excluding spaces.

ellatrix reviewed (changes requested):
Thanks for creating a PR! You say that this is to fix an issue with Chinese. In that case, the translating can indicate what needs to be counted (words or characters with or without spaces).

https://github.com/WordPress/gutenberg/blob/779a18eb2f08b2c113a54d0162bc88abccb59d7a/packages/editor/src/components/word-count/index.js#L9-L14

kjellr commented on packages/editor/src/components/table-of-contents/style.scss:
```diff
@@ -37,6 +37,12 @@
 	line-height: 30px;
 	color: $dark-gray-500;
 }
+.table-of-contents__popover .char-count {
```
Since these styles just duplicate the styles above, we can include them above like so: 

```
.table-of-contents__number,
.table-of-contents__popover .word-count,
.table-of-contents__popover .char-count {
	font-size: 21px;
	font-weight: 400;
	line-height: 30px;
	color: $dark-gray-500;
}
```

We could also probably use the more generic `.table-of-contents__number` class name instead of a new `.char-count` here. I'm not sure why Word Count uses a unique class name there. 

We should also consider adding 8px ([`$grid-size`](https://github.com/WordPress/gutenberg/blob/08bf9e1adaef7c804c4715978e79cd4af2ac8f98/assets/stylesheets/_variables.scss#L19)) bottom margin to each `table-of-contents__count` item to add a bit more vertical space between the stacked items. (We'd also want to trim the `table-of-contents__counts` container by 8px to compensate for the extra space at the bottom.)

Jackie6 replied:
Thanks for the suggestions regarding style changes. Do we also need to adjust the width of `table-of-contents__popover`, now its `min-width` is 380px. Initially, there are four columns in the popover so the width of each column is about 380/4. But now we only have three columns, so we can adjust the `min-width` to 380/4*3=285px.

kjellr replied:
Thanks, @Jackie6! This looks great. Since the character count will likely be a longer number than the others, I think it makes sense to include a little more spacing here than there was before. I agree that the current `380px` is a little wide, but something like `320px` feels just right to me: 

<img width="388" alt="Screen Shot 2019-04-04 at 8 00 51 AM" src="https://user-images.githubusercontent.com/1202812/55554035-ccfe6380-56af-11e9-81bd-1c6056ed8373.png">

Let's adjust that — once that's in place, this is all set from a design perspective. ✅ 

Thank you for all your work here!

kjellr reviewed (commented):
Thanks for clarifying about the design feedback, @talldan — I forgot to link to the discussion. 

@Jackie6 — I left a couple comments above. Aside from that, GitHub's now saying that there are 336 files changed in this PR... is that intentional? I think this may have gotten out of sync somewhere.

Jackie6 committed:
Merge branch 'master' of https://github.com/WordPress/gutenberg into update/word-count

Jackie6 committed:
Remove duplicate styles and add bottom margin

Jackie6 committed:
Adjust the width of table of contents popover

mapk reviewed (approved):
Looks great from a design perspective. Thanks @Jackie6.

![Screen Shot 2019-04-04 at 7 34 24 PM](https://user-images.githubusercontent.com/617986/55600433-cf44da00-5710-11e9-9769-756de0b030f2.png)


LGTM

talldan commented:
I think this looks good in terms of the code changes.

I'm now curious about @ellatrix's comment—it seems that the existing word count should already support Chinese characters, but the translation for that string is probably not correct (or untranslated).

Still, it could be worth adding character count as an extra feature and also look to fix the translation?

Jackie6 commented:
Hi @ellatrix thanks for your comments. 
As discussed in #13796 , chinese characters are not word-based, for example, "Hello world" has two English words, ten English characters. But in chinese, "你好世界" has four chinese characters or words. Thus, it may reqiure a major change on the `word-count` module to support counting the words in mixed languages (such as English+Chinese). We cannot merely pass `words`, `characters_excluding_spaces` or  `characters_including_spaces` to the `word-count` function. But 
> and a character count could likely be useful regardless of language.

This is the reason why this pull request is created.

talldan reviewed (approved):
#13225 is a good justification for this. Adding my approval.

ellatrix commented:
@Jackie6 So this change is meant to be for a mixed script post? It still does not detect which words/characters belong to which script? You'll just have both counters, neither of which seem useful for the post as a whole?

I'm not sure if character count is very useful in other scripts. Usually, _if_ a Latin script writer wants to know the total character count, they want to include spaces.

ellatrix commented on packages/editor/src/components/char-count/index.js:
```diff
@@ -0,0 +1,24 @@
+	 * Only count the number of characters (no spaces).
+	 * Do not translate into your own language.
+	 */
+	const charCountType = _x( 'characters_excluding_spaces', 'Word count type. Do not translate!' );
```
I see #13225. Since the default language is English, I think it's better to change this to  `characters_including_spaces`.

```suggestion
	const charCountType = _x( 'characters_including_spaces', 'Character count type. Do not translate literally!' );
```

Jackie6 replied:
Done

ellatrix commented on packages/editor/src/components/char-count/index.js:
```diff
@@ -0,0 +1,24 @@
+function CharCount( { content } ) {
+	/*
+	 * Only count the number of characters (no spaces).
+	 * Do not translate into your own language.
```
Might be good to list the two options here: `characters_excluding_spaces` and `characters_including_spaces`.

Jackie6 replied:
Done.

Jackie6 commented:
Hi @ellatrix, thanks for clarifying this pr further. This pr still does not fix the counting for mixed script post. But it can count the correct number of Chinese words. 
![image](https://user-images.githubusercontent.com/24849874/55905460-f7fd2180-5b9f-11e9-8cbd-720a62fdf6de.png)
 I expect to create a new feature to support counting the words in a mixed script post in the future. But at this point, I think it is good to add characters count in the content structure as most editors already provide this. An example is MS Word:
![image](https://user-images.githubusercontent.com/24849874/55905637-6346f380-5ba0-11e9-9ce5-72b0a3c87082.png)

Jackie6 committed:
List two options available for char-count

ellatrix commented:
@Jackie6 Thanks, this is looking great!

One final thing: for Latin scripts it might make sense to include both word count and character count, but does it also make sense for e.g. a Chinese translation? Should word count not be removed in that case (perhaps by allowing an empty translation)?

Jackie6 commented:
Hi @ellatrix thanks for comments. Do you mean we need to remove the word count when the language is Chinese? I think it will cause the inconsistency of UI. That is sometimes both word count and char count show but sometimes only char count show.

ellatrix commented:
@Jackie6 Yes, the UI will be different, but is word count useful at all in a Chinese translation? If it's not useful, it shouldn't be there?

Jackie6 commented:
@ellatrix But apart from Chinese, Korean and Japenese are also character-based rather than word-based. For languages like these, we also need to remove the word count?

dmsnell commented:
Just a random drive-by, but even though we have characters in CJK languages that by themselves are words it doesn't mean that there aren't multi-character words.

Other languages such as early Greek, Linear B, early Egyption also don't use spaces to separate words. Okay, we'll ignore the fact that these earlier languages could be equally written left-to-right, right-to-left, top-to-bottom, bottom-to-top, or changing text direction at intervals to form spirals and other patterns.

[Word segmentation](https://www.quora.com/Which-tools-do-you-recommend-for-CJK-language-word-segmentation-and-stemming) is a real problem with an extremely complicated solution. Check out this [post from the Solr project talking about ambiguity](https://www.hathitrust.org/blogs/large-scale-search/multilingual-issues-part-1-word-segmentation) in Chinese word segmentation - ambiguity that exists because there are multiple valid ways to segment the words in the sentence.

![](http://www.hathitrust.org/sites/www.hathitrust.org/files//CJK_teehan_brocolli.png)

> the first segmentation means “I like New Zealand Flowers.” The second segmentation means “I like fresh broccoli” 

---

it's impractical to think that we have the energy and time at the moment to conclusively solve this generally-unsolved problem but we do have some reasonable ways we can move forward in the meantime:

 - cut out word counts entirely: a drastic choice that excludes being helpful to a large swath of cases where it otherwise would be
 - cut out word counts for any language we think might be troublesome: this is the conservative option to prevent making basic mistakes but may not be super helpful
 - add some qualifying notice to the word count when we think it could be wrong, such as when we detect characters from languages that don't us spaces for segmentation or when the ratio of spaces to characters is below a given threshold

my shallow thoughts on the matter are that it'd be more helpful to show the word count with some disclaimer - maybe a popover tooltip explaining that we think it could be wrong - because if we hide it altogether at the first detection of a Chinese character then we'll be degrading the experience without an explanation for not only Chinese posts but also posts that quote Chinese text, even if it's just a small one.

for inclusionary purposes I think it would make more sense to develop some basic heuristics to detect when we think the word count could be wrong; possible indicators based on no research could be:
 - word-to-character ratio is out of whack, less than 1:20 (spaces to other characters)
 - we encounter a long stream of characters in non-roman letters, we could treat the blob as a whole or split it based on approximate character-per-words estimates

ellatrix commented:
Thanks @dmsnell for you comment.

My (small) problem with this change is that each of these counts can be translated.

So e.g. a Chinese translating might be:

* `words` => `characters_excluding_spaces`.
* `characters` => `characters_excluding_spaces`.

So you'd just have two counts that are the same thing.

Possible solution:

* Remove word count if `__( 'words' ) === __( 'characters' )`.
* OR remove word count if `__( 'words' ) === 'characters_excluding_spaces' && __( 'characters' ) === 'characters_excluding_spaces'`, and if `__( 'words' ) === 'characters_excluding_spaces'`, force character count to be `characters_including_spaces`, or if `__( 'words' ) === 'characters_including_spaces'`, remove character count.

What do you think @dmsnell and @Jackie6?

ellatrix commented:
Another possible solution is to display all three possible counters, and don't translate anything.

dmsnell commented:
> Remove word count if __( 'words' ) === __( 'characters' ).

@ellatrix this suggestion is much more favorable in my opinion because it's starting at least with some kind of principle: if the translator says that these are the same then only show one

special-casing specific languages in groups I'm sure will cause more stir and degradation than we intend.

having looked at `@wordpress/wordcount` it became clear that we're starting from the point of poking-at-holes vs. building a solid ground. that is, making this robustly-solved is a much bigger project than just showing characters and so I think getting character counts out even when word counts will be wrong is a positive step forward

I'm a fan of adding a disclaimer popup but I can see where that could introduce needless confusion - for the time being checking if `__( 'words' ) === __( 'characters' )` or whatever form is used here `__( 'words', { context: 'label for word-count' } )` is a good tradeoff between _getting it perfect_ and _getting it shipped_

Jackie6 commented:
@ellatrix I am a little confused. Why would `__( 'words' ) === __( 'characters' )` ? In Chinese, the translated strings of `words` and `characters` would be different. And I guess we cannot control the localization here, right? From my understanding, we just wrap translatable strings in the `__()` function, and translators will translate the strings.

ellatrix commented:
@Jackie6 If it is different, it would still be displayed. If it's translated the same, one of them is removed. It doesn't make sense to have both counts if they are the same. This may not be the case in Chinese, but maybe it is in another language.

Jackie6 committed:
Remove word count if it equals to char count

dmsnell commented on packages/editor/src/components/table-of-contents/panel.js:
```diff
@@ -21,10 +21,13 @@ function TableOfContentsPanel( { headingCount, paragraphCount, numberOfBlocks, h
-					<WordCount />
-				</div>
+				{
+					__( 'Word' ) !== __( 'Characters' ) &&
```
oops! `Word` vs. `Words`

Jackie6 replied:
Fixed. Sorry about that.

swissspidy commented on packages/editor/src/components/table-of-contents/panel.js:
```diff
@@ -20,9 +21,16 @@ function TableOfContentsPanel( { headingCount, paragraphCount, numberOfBlocks, h
 				tabIndex="0"
 			>
+				{
+					__( 'Words' ) !== __( 'Characters' ) &&
```
What is the goal of this? Do you rely on translators on changing this? That's not possible as there is no context.

ellatrix replied:
@swissspidy Ah, yes, I meant https://github.com/WordPress/gutenberg/pull/14589#issuecomment-490989857

ellatrix replied:
We should export the translation in both count components, then compare the types here. https://github.com/WordPress/gutenberg/pull/14589/files#r282803952

swissspidy commented:
@ellatrix I assume with `Remove word count if __( 'words' ) === __( 'characters' )` you meant `charCountType === wordCountType`? Because `__( 'words' )` is totally different from `_x( 'words', 'Word count type. Do not translate!' )`.

ellatrix commented on packages/editor/src/components/char-count/index.js:
```diff
@@ -0,0 +1,24 @@
+	 * Two options available for counting: 'characters_excluding_spaces' or 'characters_including_spaces'
+	 * Do not translate into your own language.
+	 */
+	const charCountType = _x( 'characters_including_spaces', 'Character count type. Do not translate literally!' );
```
It would be good to export this, as well in `word-count`, to be used `table-of-contents` to compare.

aristath commented:
If I'm not mistaken this one can be closed now that https://github.com/WordPress/gutenberg/pull/24823 was merged?

swissspidy commented:
@aristath Sort of. #24823 doesn't really fully address #13796, which this PR aimed to do.

t-hamano commented:
I tried to understand what this PR is trying to solve.

The character count is implemented by #24823. I think the rest of what this PR is trying to do is to add an option to control the presence or absence of spaces in the character count.

In word count, we can control the count type by `words`, `characters_excluding_spaces` and `characters_including_spaces`.

https://github.com/WordPress/gutenberg/blob/aa9b1d665a10f32182797e0d7e7a559362d1ec1c/packages/editor/src/components/word-count/index.js#L19-L24

In character count, on the logic is fixed on `characters_including_spaces`.

https://github.com/WordPress/gutenberg/blob/aa9b1d665a10f32182797e0d7e7a559362d1ec1c/packages/editor/src/components/character-count/index.js#L18

Is this the right goal for this PR?

youknowriad commented:
Doing some cleanup in the repository. I was wondering if this PR still worth keeping open. Probably needs to be redone entirely if we ever get to it. Let's close it. Please let me know if you think otherwise.
